### ISSUE TITLE
Timeout Exception issues with Tab's 

### ISSUE THREAD

**Issue by auphofBSF** [ClericPy/ichrome#6]
Experiencing Timeout Exception issues with Tab's 

Happens whilst running your example (as per below) and waiting more than 5 seconds to accept 
the alert

Suggest adding `websocket._exceptions.WebSocketTimeoutException` to the 'try/except' in Tab._recv_daemon https://github.com/ClericPy/ichrome/blob/749f674027801e21fdffe4472892df1f47222b3d/ichrome/sync_utils.py#L227

# TERMINAL/LOG OUTPUT
```
2020-02-16 18:53:54 {INFO } [ichrome] sync_utils.py(86): new tab Tab(https://pypi.org/)
2020-02-16 18:53:54 {INFO } [ichrome] sync_utils.py(246): <Tab(https://pypi.org/)> send: {'method': 'Page.enable', 'params': {}, 'id': 1}
2020-02-16 18:53:57 {INFO } [ichrome] sync_utils.py(246): <Tab(https://pypi.org/)> send: {'method': 'Runtime.evaluate', 'params': {'expression': 'alert("test ok")'}, 'id': 2}
2020-02-16 18:54:02 {INFO } [ichrome] sync_utils.py(246): <Tab(https://pypi.org/)> send: {'method': 'Page.close', 'params': {}, 'id': 3}
Exception in thread Thread-7:
Traceback (most recent call last):
  File "C:\Users\BSFau\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_socket.py", line 102, in recv
    bytes_ = _recv()
  File "C:\Users\BSFau\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_socket.py", line 84, in _recv
    return sock.recv(bufsize)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\threading.py", line 926, in _bootstrap_inner
    self.run()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\ichrome\sync_utils.py", line 215, in _recv_daemon
    data_str = self.ws.recv()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_core.py", line 314, in recv
    opcode, data = self.recv_data()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_core.py", line 331, in recv_data
    opcode, frame = self.recv_data_frame(control_frame)
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_core.py", line 344, in recv_data_frame
    frame = self.recv_frame()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_core.py", line 378, in recv_frame
    return self.frame_buffer.recv_frame()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_abnf.py", line 361, in recv_frame
    self.recv_header()
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_abnf.py", line 309, in recv_header
    header = self.recv_strict(2)
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_abnf.py", line 396, in recv_strict
    bytes_ = self.recv(min(16384, shortage))
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_core.py", line 453, in _recv
    return recv(self.sock, bufsize)
  File "C:\Users\<USER>\Anaconda3\envs\<ENV>\lib\site-packages\websocket\_socket.py", line 105, in recv
    raise WebSocketTimeoutException(message)
websocket._exceptions.WebSocketTimeoutException: timed out

2020-02-16 18:54:18 {INFO } [ichrome] daemon.py(308): ChromeDaemon(127.0.0.1:9222) shutting down, start-up: 2020-02-16 17:53:51, duration: 27 seconds 3 ms.
2020-02-16 18:54:18 {INFO } [ichrome] daemon.py(247): ChromeDaemon(127.0.0.1:9222) daemon break after shutdown(2020-02-16 17:54:18).
2020-02-16 18:54:18 {INFO } [ichrome] daemon.py(264): ChromeDaemon(127.0.0.1:9222) daemon exited.
```
# EXAMPLE 
```python
from ichrome import ChromeDaemon, Chrome


def main():
    with ChromeDaemon() as chromed:
        # run_forever means auto_restart
        chromed.run_forever(0)
        chrome = Chrome()
        tab = chrome.new_tab(url="https://pypi.org")
        tab.wait_loading(3)
        tab.js('alert("test ok")')
        tab.close()


if __name__ == "__main__":
    main()
```

# Environment:
Python 3.7.6 in a conda virtualenv
Windows 10 x64
platform : win-64
user-agent : conda/4.7.10 requests/2.22.0 CPython/3.6.9 Windows/10 Windows/10.0.18362
ichrome                   0.2.6                    pypi_0    pypi
websocket-client          0.57.0                   pypi_0    pypi

**Comment by ClericPy** [comment]
Here is my running logs and it runs ok. It's really hard to reproduce this Exception.

```
2020-02-16 16:22:12 {INFO } [ichrome] daemon.py(181): launch_chrome success: ChromeDaemon(127.0.0.1:9222), args: "C:/Program Files (x86)/Google/Chrome/Application/chrome.exe" --remote-debugging-address=127.0.0.1 --remote-debugging-port=9222 --user-data-dir=C:\Users\ld\ichrome_user_data\chrome_9222 "--user-agent=Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36" --disable-gpu --no-sandbox --no-first-run about:blank
2020-02-16 16:22:12 {INFO } [ichrome] sync_utils.py(84): new tab Tab(https://pypi.org/)
2020-02-16 16:22:12 {INFO } [ichrome] sync_utils.py(243): <Tab(https://pypi.org/)> send: {'method': 'Page.enable', 'params': {}, 'id': 1}
2020-02-16 16:22:14 {INFO } [ichrome] sync_utils.py(243): <Tab(https://pypi.org/)> send: {'method': 'Runtime.evaluate', 'params': {'expression': 'alert("test ok")'}, 'id': 2}
2020-02-16 16:22:19 {INFO } [ichrome] sync_utils.py(243): <Tab(https://pypi.org/)> send: {'method': 'Page.close', 'params': {}, 'id': 3}
2020-02-16 16:22:19 {INFO } [ichrome] daemon.py(308): ChromeDaemon(127.0.0.1:9222) shutting down, start-up: 2020-02-16 16:22:10, duration: 9 seconds 604 ms.
2020-02-16 16:22:19 {INFO } [ichrome] daemon.py(247): ChromeDaemon(127.0.0.1:9222) daemon break after shutdown(2020-02-16 16:22:19).
2020-02-16 16:22:19 {INFO } [ichrome] daemon.py(264): ChromeDaemon(127.0.0.1:9222) daemon exited.
```

Test env: Windows10, python3.7 without conda, Chrome 80.

It seems that in your testing, there is a crash for websocket connection/recv failure. The ws connection is a runtime level requirement for all the communication between the two endpoints, which means all the features to control chrome instance can not make sense without the websocket connection. To handle this timeout error with continure or break will make a mistake.

Here is my 3 opinions:

1. would you mind to give a try for async version's ichrome instead of the sync version.
2. Ensure your permission to create ws for this 9222 port, and check the network.
3. Is the timeout exception a frequently occurring issue, or only sometimes?

### To check your websocket:

```python
from ichrome import ChromeDaemon
import requests
import json
import websocket


def main():
    with ChromeDaemon(port=9222) as chromed:
        chromed.run_forever(0)
        r = requests.get('http://127.0.0.1:9222/json')
        webSocketDebuggerUrl = r.json()[0]['webSocketDebuggerUrl']
        ws = websocket.WebSocket()
        ws.connect(webSocketDebuggerUrl, timeout=5)
        ws.send(
            json.dumps({
                "method": "Runtime.evaluate",
                "params": {
                    "expression": "window.location.href"
                },
                "id": 1
            }))
        print(ws.recv())
        # {"id":1,"result":{"result":{"type":"string","value":"about:blank"}}}


if __name__ == "__main__":
    main()
```

Ensure that print log is {"id":1,"result":{"result":{"type":"string","value":"about:blank"}}}.

### Or try the async version

```python
from ichrome import ChromeDaemon, AsyncChrome
import asyncio


async def main():
    with ChromeDaemon(port=9222) as chromed:
        chromed.run_forever(0)
        # `async with` usage will be fixed after version 0.2.7
        chrome = AsyncChrome(port=9222)
        await chrome.connect()
        tab = (await chrome.tabs)[0]
        async with tab():
            print(await tab.current_url)
            # about:blank
        await chrome.close()


if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by ClericPy** [comment]
Thanks for your help 😃
